### PR TITLE
Release 3.0

### PR DESCRIPTION
**Pull request by hwillson**
This release PR will serve to collect significant new features, deprecation warnings, and minor breaking changes that we intend to release in `graphql-subscriptions@3.0.0`. It should not be merged until we're ready to release 3.0.

### CODE REVIEW

**Comment by maxpain** [comment]
Any updates?

**Comment by Narretz** [comment]
Are there any more breaking changes planned? If not, could the release happen? Or even if, could the release happen? There's always the next major version for more breaking changes.

**Comment by daphnesmit** [comment]
Same 👯

**Comment by Hetch3t** [comment]
Any new improvements planned? What's holding you from merging it finally?

**Comment by maxpain** [comment]
Any news?

**Comment by Adel-ak** [comment]
1 year and still not ready? not to be rude or anything, what's the hold-up

**Comment by aidenfoxx** [comment]
Any update would be appreciated? :pray:

**Comment by MrDesjardins** [comment]
Hello,

I am reading the latest (V4) [Apollo documentation](https://www.apollographql.com/docs/apollo-server/data/subscriptions/) that recommend using the `graphql-subscriptions` library with subscription. However, the type, as mentioned by many people in this thread, are not compatible. I believe this pull request will solve this issue. Apollo is quite a popular and the are fading out their [v3](https://www.npmjs.com/package/apollo-server) for [v4](https://www.npmjs.com/package/@apollo/server).

There is a [discussion in the graphql-codegen](https://github.com/dotansimha/graphql-code-generator/issues/7197) that generate code to use with Apollo that highlight that the issue is not on the generated type. Someone in the Codegen discussion gives a quick hack to fix the issue: 
```
subscribe: () => ({
        [Symbol.asyncIterator]: () => pubsub.asyncIterator('NUMBER_INCREMENTED'),
      }),
```
However, the hack is only for type. Doing:

```
subscribe: () => pubsub.asyncIterator("NUMBER_INCREMENTED") as any,
```
Also fix the problem. 

The first comment of this thread is from `Nov 25, 2021` and we are now almost over with 2022. The direction of this library is unclear to me but it is an helpful one and it would be great to see it moving forward to continue to work with TypeScript without hacks. Have a great day.

File: src/with-filter.ts [hunk 1]
```diff
@@ -1,10 +1,8 @@
-import { $$asyncIterator } from 'iterall';
-
 export type FilterFn<TSource = any, TArgs = any, TContext = any> = (rootValue?: TSource, args?: TArgs, context?: TContext, info?: any) => boolean | Promise<boolean>;
-export type ResolverFn<TSource = any, TArgs = any, TContext = any> = (rootValue?: TSource, args?: TArgs, context?: TContext, info?: any) => AsyncIterator<any>;
+export type ResolverFn<TSource = any, TArgs = any, TContext = any> = (rootValue?: TSource, args?: TArgs, context?: TContext, info?: any) => AsyncIterator<any> | Promise<AsyncIterator<any>>;
```
Comment (Cellule) [inline src/with-filter.ts:2]: I think to have proper typing there needs to be a distinction between the `asyncIteratorFn` received as input which needs to return an `AsyncIterator` and the `withFilter`method that actually returns a `AsyncIterableIterator`

Should split `ResolverFn` in 2 types to make the distinction.

**Comment by Aeolun** [comment]
@hwillson Is this something that is still on the agenda? Migrating to apollo v4 would be a lot more pleasant without conflcting types.

File: src/with-filter.ts (hunk 1 above)
Comment (tninesling) [reply src/with-filter.ts:2]: Added in #276

**Review by tninesling — APPROVED** [review]

**Comment by chdanielmueller** [comment]
Hi @tninesling 
Great to see this merged. Is there a timeline when we can expect this version to be released to npm?
Thanks

**Pull request merged**